lib/game/Unification.ts

```diff
@@ -29,60 +30,78 @@ function unifyVariable(currentAssignment: Assignment, v: VariableName, term: Ter
-        } else {
-            return unifyEquation(currentAssignment, [v1Value, { variable: v2 }])
-        }
-    } else {
-        if (currentAssignment.isAssigned(v2)) {
-            const v2Value = currentAssignment.getAssignedValue(v2)!
-            return unifyEquation(currentAssignment, [{ variable: v1 }, v2Value])
-        } else {
-            currentAssignment.unite(v1, v2)
-            return true
-        }
-    }
+    let v1Value = currentAssignment.getAssignedValue(v1);
+    let v2Value = currentAssignment.getAssignedValue(v2);
+
+    if (v1Value !== undefined && v2Value !== undefined)
+        return unifyEquation(currentAssignment, [v1Value, v2Value]); 
+
+    if (v1Value !== undefined)
+        return unifyEquation(currentAssignment, [v1Value, { variable: v2 }]);
+
+    if (v2Value !== undefined)
+        return unifyEquation(currentAssignment, [{ variable: v1 }, v2Value]);
+
+    currentAssignment.unite(v1, v2);
+    return true;
 }
 
-function unifyEquation(currentAssignment: Assignment, equation: Equation): boolean {
+function unifyEquation(currentAssignment: Assignment, equation: TermEquation): boolean {
     const [lhs, rhs] = equation
-    if ("variable" in lhs) {
-        if ("variable" in rhs) {
-            return unifyVariables(currentAssignment, lhs.variable, rhs.variable)
-        } else {
-            return unifyVariable(currentAssignment, lhs.variable, rhs)
-        }
-    } else {
-        if ("variable" in rhs) {
-            return unifyVariable(currentAssignment, rhs.variable, lhs)
-        } else {
-            if (lhs.label !== rhs.label) {
-                return false
-            }
-            if (lhs.args.length !== rhs.args.length) {
-                return false
-            }
-            let unifiedSuccessfully: boolean[] = []
-            for (let i = 0; i < lhs.args.length; i++) {
-                const lhsArg = lhs.args[i]
-                const rhsArg = rhs.args[i]
-                unifiedSuccessfully.push(unifyEquation(currentAssignment, [lhsArg, rhsArg]))
-            }
-            return (!unifiedSuccessfully.includes(false))
-        }
+
+    if ("variable" in lhs && "variable" in rhs)
+        return unifyVariables(currentAssignment, lhs.variable, rhs.variable);
+
+    if ("variable" in lhs)
+        return unifyVariable(currentAssignment, lhs.variable, rhs);
+
+    if ("variable" in rhs)
+        return unifyVariable(currentAssignment, rhs.variable, lhs);   
+
+    if ("number" in lhs)
+        return ("number" in rhs) && lhs.number === rhs.number;
+
+    if ("number" in rhs) return false;
+
+    // Now, lhs and rhs must be function terms
+    if (lhs.function !== rhs.function) return false;
+    if (lhs.args.length !== rhs.args.length) return false;
+
+    for (let i = 0; i < lhs.args.length; i++) {
+        const lhsArg = lhs.args[i]
+        const rhsArg = rhs.args[i]
+        if (!unifyEquation(currentAssignment, [lhsArg, rhsArg]))
+            return false;
     }
+
+    return true;
```

Comment:
This change is undesirable, as we still wish for unification to continue after failure.